test/io/client_test.dart

```diff
@@ -132,4 +146,22 @@ void main() {
 
     expect(socket, isNotNull);
   });
+
+  test('runWithClient', () {
+    late http.Client client;
+    http.runWithClient(() => client = http.Client(), () => TestClient());
```

Comment:
Can just do
```dart
 http.Client client = http.runWithClient(() => http.Client(), TestClient.new);
```
since you return the value anyway.
(Can also use a constructor tear-off for that, but I don't find that actually more readable:
```dart
 http.Client client = http.runWithClient(http.Client.new, TestClient.new);
```
)

Reply:
Done (without the tear-offs since they aren't available in Dart 2.14)